src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentResult.cs

```diff
@@ -0,0 +1,21 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.CodeAnalysis.Rename
+{
+    public static partial class Renamer
+    {
+        public struct RenameDocumentResult
+        {
+            internal RenameDocumentResult(Solution solution, bool renameSymbolSucceeded)
+            {
+                this.Solution = solution;
+                this.RenameSymbolSucceeded = renameSymbolSucceeded;
+            }
+
+            public Solution Solution { get; }
+            public bool RenameSymbolSucceeded { get; }
```

Comment:
if this fails, is there no additional information we can give as to why?  For example, internally, several APIs return a bool and a message (which is intended to be user visible).

Reply:
As a public API, i would like this to be documented as part of this PR.

Reply:
Is message + bool (or even an enum?) the right approach here? I'm open to suggestions.

Reply:
Proposal for modified result to get conversation going

```csharp
public class RenameDocumentResult
{
    public Solution Solution { get; }
    public RenameActionResult TypeRenamedResult { get; }
    public RenameActionResult NamespaceSyncResult { get; }
    public string Message { get;  }
}

public enum RenameActionResult
{
    NoOperation,
    Succeeded,
    Failed
}
```

Reply:
So the goal here to let the caller (in our prototypical case, the project system) know if:

1. We did nothing at all.
2. We ran into errors of some kind.

If there are errors, what is the expectation? Are we going to show individual errors? Or as long as one thing succeeded, just go with that?

It seems like this API has some challenges around aggregation of the problems: if we want to show all the errors, the caller has to go to each Result member and aggregate it. And if we add a new one (since this is going to update file headers, right?) then we have to go back to callers to update that logic. I'm not sure if just `List<string> Errors` seems easier, and you know there weren't errors if the list is empty.

Reply:
> since this is going to update file headers, right?

Hopefully :) 

> It seems like this API has some challenges around aggregation of the problems: if we want to show all the errors, the caller has to go to each Result member and aggregate it.

If we stop on first error, one message will do. Otherwise, I agree that we should have some aggregation of error messages. I don't particularly want to bail just because on action potentially failed. Instead, the caller can decide on the behavior for that.

```csharp
public class RenameDocumentResult
{
    public Solution Solution { get; }
    public RenameActionResult Result { get; }
    public ImmutableArray<string> Errors { get;  }
}

public enum RenameActionResult
{
    NoOperation,
    Succeeded,
    Failed
}
```

Reply:
https://github.com/dotnet/roslyn/pull/41267#discussion_r385988467 has good discussion that I think warrants a tweak on this. 

We could have something like below 

```csharp 
public class RenameDocumentResult
{
    public Solution Solution { get; }
    public RenameActionResult Result { get; }
    public ImmutableArray<string> Errors { get;  }
}

public class RenameDocumentAction
{
    // Message describing what the action is doing
    public string Message { get; }

    // Perform the action and update the solution
    public Task<RenameDocumentResult> UpdateSolutionAsync(Solution solution)
}

public enum RenameActionResult
{
    NoOperation,
    Succeeded,
    Failed
}

class Renamer 
{
    ImmutableArray<RenameDocumentAction> RenameDocument(...);
}
```

Reply:
@ocallesp @davkean could you way in on the usefulness of some of this behavior?

Reply:
I'd imagine that some expression of "no operation" would appear in RenameDocumentAction of some kind, because effectively we need to know whether we need to pop a dialog quickly. The UpdateSolutionAsync is doing the real slow bit of actually updating files, but the "is there a type we can rename" is something done up front. Ditto for checking namespaces or file headers.

Reply:
There's two parts: 

1. The cheap "Should we try to make this change" part, where it generates a RenameDocumentAction.

2. The expensive "Try to apply this change" part, where actual changes may or may not have been applied. Maybe it's not important to know if the solution is different from before.

The main part is that an action would not be generated at all if we can detect it's not needed.

Reply:
I've updated the main description with the new API shape. I'm going to resolve this conversation. We should start a new thread since the design is drastically different than the original